src/grpc/packet_router/hpr_packet_service.erl

```diff
@@ -9,32 +9,112 @@
 ]).
 
 -export([
-    packet_down/2
+    send_downlink/2
 ]).
 
+-define(REG_KEY(Gateway), {?MODULE, Gateway}).
+
 -spec init(atom(), StreamState :: grpcbox_stream:t()) -> grpcbox_stream:t().
 init(_RPC, StreamState) ->
     StreamState.
 
--spec route(hpr_packet_up:packet(), grpcbox_stream:t()) ->
-    {ok, grpcbox_stream:t()} | grpcbox_stream:grpc_error_response().
-route(PacketUp, StreamState) ->
+-spec route(hpr_envelope_up:envelope(), grpcbox_stream:t()) ->
+    {ok, grpcbox_stream:t()} | {stop, grpcbox_stream:t()}.
+route(EnvUp, StreamState) ->
     Self = self(),
-    _ = erlang:spawn(hpr_routing, handle_packet, [PacketUp, Self]),
-    {ok, StreamState}.
+    case hpr_envelope_up:data(EnvUp) of
+        {packet, PacketUp} ->
+            _ = erlang:spawn(hpr_routing, handle_packet, [PacketUp, Self]),
+            {ok, StreamState};
+        {register, Reg} ->
+            Gateway = hpr_register:gateway(Reg),
+            case hpr_register:verify(Reg) of
+                false ->
+                    lager:info("failed to verify"),
+                    {stop, StreamState};
+                true ->
+                    true = gproc:add_local_name(?REG_KEY(Gateway)),
+                    {ok, StreamState}
+            end
+    end.
 
 -spec handle_info(Msg :: any(), StreamState :: grpcbox_stream:t()) -> grpcbox_stream:t().
-handle_info({packet_down, PacketDown}, StreamState) ->
-    grpcbox_stream:send(false, PacketDown, StreamState);
+handle_info({downlink, PacketDown}, StreamState) ->
+    EnvDown = hpr_envelope_down:new(PacketDown),
+    grpcbox_stream:send(false, EnvDown, StreamState);
 handle_info(_Msg, StreamState) ->
     StreamState.
 
--spec packet_down(Pid :: pid(), PacketDown :: hpr_packet_down:packet()) -> ok.
-packet_down(Pid, PacketDown) ->
+-spec send_downlink(Pid :: pid(), EnvDown :: hpr_envelope_down:packet()) -> ok.
+send_downlink(Pid, EnvDown) ->
     case erlang:is_process_alive(Pid) of
         true ->
-            Pid ! {packet_down, PacketDown};
+            Pid ! {downlink, EnvDown};
         false ->
-            lager:warning("failed to send packet_down to stream ~p", [Pid])
+            lager:warning("failed to send envelope_down to stream ~p", [Pid])
     end,
     ok.
+
+%% ------------------------------------------------------------------
+%% EUnit tests
+%% ------------------------------------------------------------------
+-ifdef(TEST).
+
+-include_lib("eunit/include/eunit.hrl").
+
+init_test() ->
+    ?assertEqual(stream_state, ?MODULE:init(rpc, stream_state)),
+    ok.
+
+route_packet_test() ->
+    meck:new(hpr_routing, [passthrough]),
+    Self = self(),
+    PacketUp = hpr_packet_up:new(#{}),
+    EnvUp = hpr_envelope_up:new(PacketUp),
+    meck:expect(hpr_routing, handle_packet, [PacketUp, Self], ok),
+
+    ?assertEqual({ok, stream_state}, ?MODULE:route(EnvUp, stream_state)),
+
+    ?assertEqual(1, meck:num_calls(hpr_routing, handle_packet, 2)),
+
+    meck:unload(hpr_routing),
```

Comment:
It might be better to make a `setup` fixture and use `meck:unload` in the per test cleanup. That way if the test asserts we aren't left with the mock in subsequent tests.